lms/djangoapps/commerce/utils.py

```diff
@@ -109,3 +117,247 @@ def upgrade_url(self, user, course_key):
+
+        _process_refund(
+            refund_ids=refund_ids,
+            api_client=api_client,
+            course_product=course_enrollment,
+        )
+    else:
+        log.info('No refund opened for user [%s], course [%s]', enrollee.id, course_key_str)
+
+    return refund_ids
+
+
+def _process_refund(refund_ids, api_client, course_product):
+    """
+    Helper method to process a refund for a given course_product
+
+    Returns:
+        bool: True if the refund process was successful, False if there are any Errors that are not handled
+    """
+    config = CommerceConfiguration.current()
+
+    if config.enable_automatic_refund_approval:
+        refunds_requiring_approval = []
+
+        for refund_id in refund_ids:
+            try:
+                # NOTE: Approve payment only because the user has already been unenrolled. Additionally, this
+                # ensures we don't tie up an additional web worker when the E-Commerce Service tries to unenroll
+                # the learner
+                api_client.refunds(refund_id).process.put({'action': 'approve_payment_only'})
+                log.info('Refund [%d] successfully approved.', refund_id)
+            except:  # pylint: disable=bare-except
+                # Push the refund to Support to process
+                log.exception('Failed to automatically approve refund [%d]!', refund_id)
+                refunds_requiring_approval.append(refund_id)
+    else:
+        refunds_requiring_approval = refund_ids
+
+    if refunds_requiring_approval:
+        # XCOM-371: this is a temporary measure to suppress refund-related email
+        # notifications to students and support for free enrollments.  This
+        # condition should be removed when the CourseEnrollment.refundable() logic
+        # is updated to be more correct, or when we implement better handling (and
+        # notifications) in Otto for handling reversal of $0 transactions.
+        if course_product.mode != 'verified':
+            # 'verified' is the only enrollment mode that should presently
+            # result in opening a refund request.
+            msg = 'Skipping refund support notification for non-verified mode for user [%s], course [%s], mode: [%s]'
+            course_identifier = course_product.course_id
+            if course_product.uuid:
+                course_identifier = str(course_product.uuid)
+                msg = ('Skipping refund support notification for non-verified mode for user [%s], '
+                       'course entitlement [%s], mode: [%s]')
+            log.info(
+                msg,
+                course_product.user.id,
+                course_identifier,
+                course_product.mode,
+            )
+            return False
+        else:
+            try:
+                _send_refund_notification(course_product, refunds_requiring_approval)
```

Comment:
This method contains many try/catch blocks and doesn't seem to allow exceptions to bubble up to this point. If Zendesk is down or something goes wrong creating the tickets, we may not know and may end up returning True from this method, indicating that it's safe to move ahead with revoke/unenroll.